code/modules/mob/living/basic/space_fauna/revenant/_revenant.dm

```diff
@@ -0,0 +1,499 @@
+	/// The icon we use when stunned (temporarily frozen)
+	var/icon_stun = "revenant_stun"
+	/// The icon we use while draining someone.
+	var/icon_drain = "revenant_draining"
+
+	/// Are we currently dormant (ectoplasm'd)?
+	var/dormant = FALSE
+	/// If the revenant can take damage from normal sources.
+	var/revealed = FALSE
+	/// If the revenant's abilities are blocked by a holy power.
+	var/inhibited = FALSE
+	/// Are we currently draining someone?
+	var/draining = FALSE
+	/// Have we already given this revenant abilities?
+	var/generated_objectives_and_spells = FALSE
+
+	/// Lazylist of drained mobs to ensure that we don't steal a soul from someone twice
+	var/list/drained_mobs = null
+	/// List of action ability datums to grant on Initialize. Keep in mind that anything with the `/aoe/revenant` subtype starts locked by default.
+	var/static/list/datum/action/abilities = list(
+		/datum/action/cooldown/spell/aoe/revenant/blight,
+		/datum/action/cooldown/spell/aoe/revenant/defile,
+		/datum/action/cooldown/spell/aoe/revenant/haunt_object,
+		/datum/action/cooldown/spell/aoe/revenant/malfunction,
+		/datum/action/cooldown/spell/aoe/revenant/overload,
+		/datum/action/cooldown/spell/list_target/telepathy/revenant,
+	)
+
+	/// The resource, and health, of revenants.
+	var/essence = 75
+	/// The regeneration cap of essence (go figure); regenerates every Life() tick up to this amount.
+	var/max_essence = 75
+	/// If the revenant regenerates essence or not
+	var/essence_regenerating = TRUE
+	/// How much essence regenerates per second
+	var/essence_regen_amount = 2.5
+	/// How much essence the revenant has stolen
+	var/essence_accumulated = 0
+	/// How much stolen essence is available for unlocks
+	var/essence_excess = 0
+	/// How long the revenant is revealed for, is about 2 seconds times this var.
+	var/unreveal_time = 0
+	/// How long the revenant is stunned for, is about 2 seconds times this var.
+	var/unstun_time = 0
+	/// How many perfect, regen-cap increasing souls the revenant has. //TODO, add objective for getting a perfect soul(s?)
+	var/perfectsouls = 0
+
+/mob/living/basic/revenant/Initialize(mapload)
+	. = ..()
+	AddElement(/datum/element/simple_flying)
+	add_traits(list(TRAIT_SPACEWALK, TRAIT_SIXTHSENSE, TRAIT_FREE_HYPERSPACE_MOVEMENT), INNATE_TRAIT)
+
+	for(var/ability as anything in abilities)
+		var/datum/action/spell = new ability(src)
+		spell.Grant(src)
+
+	RegisterSignal(src, COMSIG_LIVING_BANED, PROC_REF(on_baned))
+	RegisterSignal(src, COMSIG_MOVABLE_PRE_MOVE, PROC_REF(on_move))
+	RegisterSignal(src, COMSIG_LIVING_LIFE, PROC_REF(on_life))
+	set_random_revenant_name()
+
+	GLOB.revenant_relay_mobs |= src
+
+/mob/living/basic/revenant/Destroy()
+	GLOB.revenant_relay_mobs -= src
+	return ..()
+
+/mob/living/basic/revenant/Login()
+	. = ..()
+	if(!. || isnull(client))
+		return FALSE
+
+	var/static/cached_string = null
+	if(isnull(cached_string))
+		cached_string = examine_block(jointext(create_login_string(), "\n"))
```

Comment:
do we really need to cache this tho?

Reply:
i will be spawning seven trillion revenants